File: ext/microsandbox/src/snapshot.rs
//! Snapshot management: `Microsandbox::Native::Snapshot`.
//!
//! Snapshots capture a stopped sandbox's disk state into a portable artifact
//! that a later `Sandbox.create(from_snapshot:)` can boot from. Exposed as
//! singleton functions returning plain Hashes/Arrays (shaped into value objects
//! by the Ruby layer) — there is no long-lived handle to own.
//!
//! v0.6.7 descriptor contract: an artifact is identified by its own `name`
//! (`Snapshot::builder(name).from_sandbox(src)`), lives at `dest_dir/<name>`,
//! and its descriptor file is `snapshot.json`. `manifest.json` artifacts from
//! v0.6.6 are auto-migrated by the core on first backend connect.

use std::path::{Path, PathBuf};

use magnus::{function, prelude::*, Error, RArray, RHash, RModule, Ruby};
use microsandbox::snapshot::{
    SaveOpts, Snapshot, SnapshotFormat, SnapshotHandle, SnapshotScope, SnapshotVerifyReport,
    UpperVerifyStatus,
};

use crate::conv;
use crate::error;
use crate::runtime::{block_on, ruby};

fn format_str(format: SnapshotFormat) -> &'static str {
    match format {
        SnapshotFormat::Raw => "raw",
        SnapshotFormat::Qcow2 => "qcow2",
    }
}

fn scope_str(scope: SnapshotScope) -> &'static str {
    match scope {
        SnapshotScope::Disk => "disk",
        SnapshotScope::Resumable => "resumable",
    }
}

/// Parse a manifest's RFC 3339 `created_at` into epoch-ms (nil if unparseable).
fn created_at_ms(rfc3339: &str) -> Option<i64> {
    chrono::DateTime::parse_from_rfc3339(rfc3339)
        .ok()
        .map(|dt| dt.timestamp_millis())
}

/// Convert a fully-opened `Snapshot` into the `SnapshotInfo` Hash. Unlike a
/// `SnapshotHandle` (a lightweight index row), an opened snapshot carries the
/// full descriptor, so this is the richest shape — `create`/`open`/`list_dir`
/// and the `SandboxHandle#snapshot` shortcut all funnel here. `format`/
/// `fstype`/`size_bytes` are nil for checkpoint-state snapshots, and the
/// `checkpoint_*` keys are nil for file-state ones — the two state families
/// are disjoint by contract.
pub(crate) fn snapshot_to_hash(snap: &Snapshot) -> RHash {
    let m = snap.manifest();
    let state = snap.state();
    let file = state.as_file();
    let checkpoint = state.as_checkpoint();
    let hash = ruby().hash_new();
    let _ = hash.aset("digest", snap.digest().to_string());
    let _ = hash.aset("path", snap.path().to_string_lossy().into_owned());
    let _ = hash.aset("size_bytes", snap.size_bytes());
    let _ = hash.aset("scope", scope_str(m.scope));
    let _ = hash.aset("state_kind", state.kind());
    let _ = hash.aset("image_ref", m.image.reference.clone());
    let _ = hash.aset("image_manifest_digest", m.image.manifest_digest.clone());
    let _ = hash.aset("format", file.map(|f| format_str(f.format)));
    let _ = hash.aset("fstype", file.map(|f| f.fstype.clone()));
    let _ = hash.aset("checkpoint_id", checkpoint.map(|c| c.checkpoint_id.clone()));
    let _ = hash.aset(
        "checkpoint_manifest_digest",
        checkpoint.map(|c| c.manifest.clone()),
    );
    let _ = hash.aset("parent_digest", m.parent.clone());
    let _ = hash.aset("created_at_ms", created_at_ms(&m.created_at));
    let _ = hash.aset("source_sandbox", m.source_sandbox.clone());
    let labels = ruby().hash_new();
    for (k, v) in &m.labels {
        let _ = labels.aset(k.as_str(), v.as_str());
    }
    let _ = hash.aset("labels", labels);
    hash
}

/// Create a snapshot artifact named `name` from a stopped sandbox. `opts`:
/// from_sandbox (required), dest_dir, labels, force, record_integrity,
/// resumable. Returns the full SnapshotInfo Hash.
fn create(name: String, opts: RHash) -> Result<RHash, Error> {
    let mut b = Snapshot::builder(name);
    match conv::opt_string(opts, "from_sandbox")? {
        Some(src) => b = b.from_sandbox(src),
        None => {
            return Err(error::base_error(
                "snapshot create needs from_sandbox: (the source sandbox name)",
            ));
        }
    }
    if let Some(dir) = conv::opt_string(opts, "dest_dir")? {
        b = b.dest_dir(PathBuf::from(dir));
    }
    for (k, v) in conv::opt_string_map(opts, "labels")? {
        b = b.label(k, v);
    }
    if conv::opt_bool(opts, "force")? {
        b = b.force();
    }
    if conv::opt_bool(opts, "record_integrity")? {
        b = b.record_integrity();
    }
    if conv::opt_bool(opts, "resumable")? {
        b = b.resumable();
    }

    let snap = block_on(b.create()).map_err(error::to_ruby)?;
    Ok(snapshot_to_hash(&snap))
}

/// Open an existing snapshot artifact by bare name or path. Cheap metadata
/// validation only (does not read the upper file). Returns a full SnapshotInfo
/// Hash — the only way to inspect an artifact addressed by path (`get`/`list`
/// read the local index, which path-addressed artifacts are absent from).
fn open(path_or_name: String) -> Result<RHash, Error> {
    let snap = block_on(Snapshot::open(&path_or_name)).map_err(error::to_ruby)?;
    Ok(snapshot_to_hash(&snap))
}

/// Walk `dir` and parse each subdirectory's `snapshot.json` without touching
/// the local index — for enumerating external/un-imported snapshot collections.
fn list_dir(dir: String) -> Result<RArray, Error> {
    let snaps = block_on(Snapshot::list_dir(Path::new(&dir))).map_err(error::to_ruby)?;
    let arr = ruby().ary_new();
    for snap in &snaps {
        arr.push(snapshot_to_hash(snap))?;
    }
    Ok(arr)
}

/// Rebuild the local snapshot index from `dir` (defaults to the configured
/// snapshots directory). Returns the number of indexed snapshots — the repair
/// for index drift or out-of-band imports that `get`/`list` can't see.
fn reindex(dir: Option<String>) -> Result<u64, Error> {
    let dir: PathBuf = match dir {
        Some(d) => PathBuf::from(d),
        None => microsandbox::default_backend()
            .as_local()
            .map(|l| l.snapshots_dir())
            .unwrap_or_else(|| PathBuf::from(".")),
    };
    let n = block_on(Snapshot::reindex(&dir)).map_err(error::to_ruby)?;
    Ok(n as u64)
}

/// Metadata for one snapshot by name or digest.
fn get(name_or_digest: String) -> Result<RHash, Error> {
    let handle = block_on(Snapshot::get(&name_or_digest)).map_err(error::to_ruby)?;
    Ok(handle_to_hash(&handle))
}

/// All snapshots as metadata hashes.
fn list() -> Result<RArray, Error> {
    let handles = block_on(Snapshot::list()).map_err(error::to_ruby)?;
    let arr = ruby().ary_new();
    for h in &handles {
        arr.push(handle_to_hash(h))?;
    }
    Ok(arr)
}

/// Remove a snapshot artifact by name or path.
fn remove(name_or_path: String, force: bool) -> Result<(), Error> {
    block_on(Snapshot::remove(&name_or_path, force)).map_err(error::to_ruby)
}

/// Verify a snapshot's recorded upper-layer integrity. Returns
/// {digest, path, upper_status, upper_algorithm, upper_digest}. Schema-1
/// descriptors always record integrity, so the status is always "verified"
/// on success (mismatches raise SnapshotIntegrityError instead).
fn verify(name_or_path: String) -> Result<RHash, Error> {
    let snap = block_on(Snapshot::open(&name_or_path)).map_err(error::to_ruby)?;
    let report = block_on(snap.verify()).map_err(error::to_ruby)?;
    Ok(verify_report_to_hash(&report))
}

/// Bundle a snapshot into a `.tar.zst` (or plain `.tar`) archive. `opts`:
/// with_parents, with_image, plain_tar.
fn save(name_or_path: String, out_path: String, opts: RHash) -> Result<(), Error> {
    let save_opts = SaveOpts {
        with_parents: conv::opt_bool(opts, "with_parents")?,
        with_image: conv::opt_bool(opts, "with_image")?,
        plain_tar: conv::opt_bool(opts, "plain_tar")?,
    };
    block_on(Snapshot::save(
        &name_or_path,
        std::path::Path::new(&out_path),
        save_opts,
    ))
    .map_err(error::to_ruby)
}

/// Unpack a snapshot archive into the snapshots dir. Returns the loaded
/// snapshot's metadata hash. `dest` is an optional explicit directory.
fn load(archive_path: String, dest: Option<String>) -> Result<RHash, Error> {
    let dest_path = dest.map(PathBuf::from);
    let handle = block_on(Snapshot::load(
        std::path::Path::new(&archive_path),
        dest_path.as_deref(),
    ))
    .map_err(error::to_ruby)?;
    Ok(handle_to_hash(&handle))
}

fn handle_to_hash(handle: &SnapshotHandle) -> RHash {
    let hash = ruby().hash_new();
    let _ = hash.aset("digest", handle.digest().to_string());
    let _ = hash.aset("name", handle.name().map(str::to_string));
    let _ = hash.aset("parent_digest", handle.parent_digest().map(str::to_string));
    let _ = hash.aset("scope", scope_str(handle.scope()));
    let _ = hash.aset("state_kind", handle.state_kind().to_string());
    let _ = hash.aset("image_ref", handle.image_ref().to_string());
    let _ = hash.aset("format", handle.format().map(format_str));
    let _ = hash.aset("fstype", handle.fstype().map(str::to_string));
    let _ = hash.aset(
        "checkpoint_manifest_digest",
        handle.checkpoint_manifest_digest().map(str::to_string),
    );
    let _ = hash.aset("size_bytes", handle.size_bytes());
    let _ = hash.aset("locality", handle.locality().to_string());
    let _ = hash.aset("availability", handle.availability().to_string());
    let _ = hash.aset("migration_state", handle.migration_state().to_string());
    let _ = hash.aset(
        "migration_error_code",
        handle.migration_error_code().map(str::to_string),
    );
    let _ = hash.aset("path", handle.path().to_string_lossy().into_owned());
    let _ = hash.aset(
        "created_at_ms",
        handle.created_at().and_utc().timestamp_millis(),
    );
    hash
}

fn verify_report_to_hash(report: &SnapshotVerifyReport) -> RHash {
    let hash = ruby().hash_new();
    let _ = hash.aset("digest", report.digest.clone());
    let _ = hash.aset("path", report.path.to_string_lossy().into_owned());
    match &report.upper {
        UpperVerifyStatus::Verified { algorithm, digest } => {
            let _ = hash.aset("upper_status", "verified");
            let _ = hash.aset("upper_algorithm", algorithm.clone());
            let _ = hash.aset("upper_digest", digest.clone());
        }
    }
    hash
}

pub fn define(ruby: &Ruby, native: &RModule) -> Result<(), Error> {
    let class = native.define_class("Snapshot", ruby.class_object())?;
    class.define_singleton_method("create", function!(create, 2))?;
    class.define_singleton_method("open", function!(open, 1))?;
    class.define_singleton_method("get", function!(get, 1))?;
    class.define_singleton_method("list", function!(list, 0))?;
    class.define_singleton_method("list_dir", function!(list_dir, 1))?;
    class.define_singleton_method("reindex", function!(reindex, 1))?;
    class.define_singleton_method("remove", function!(remove, 2))?;
    class.define_singleton_method("verify", function!(verify, 1))?;
    class.define_singleton_method("save", function!(save, 3))?;
    class.define_singleton_method("load", function!(load, 2))?;
    Ok(())
}

File: lib/microsandbox/snapshot.rb
# frozen_string_literal: true

module Microsandbox
  # Metadata for a snapshot artifact, returned by {Snapshot.create}/{Snapshot.open}/
  # {Snapshot.get}/{Snapshot.list}/{Snapshot.list_dir}/{Snapshot.load}.
  #
  # `digest` and `path` are always present. The artifact-opening paths
  # (`create`/`open`/`list_dir`, and {SandboxHandle#snapshot}) carry the full
  # descriptor — `size_bytes`, `image_ref`, `image_manifest_digest`, `format`,
  # `fstype`, `parent_digest`, `created_at`, `source_sandbox`, and `labels`. The
  # index paths (`get`/`list`/`load`) populate `name`, `parent_digest`,
  # `image_ref`, `format`, `size_bytes`, `created_at`, and the index-only
  # `locality`/`availability`/`migration_state` columns (descriptor-only fields
  # such as `source_sandbox`/`labels` are nil/empty there).
  #
  # Snapshots come in two state families (see {#state_kind}): `"file"` — a
  # concrete disk payload, where {#format}/{#fstype}/{#size_bytes} are set — and
  # `"checkpoint"` — a manifest-backed state, where those are nil and the
  # `checkpoint_*` fields are set instead.
  class SnapshotInfo
    # @return [String] descriptor digest ("sha256:…") — the canonical identity
    attr_reader :digest
    # @return [String] artifact directory path
    attr_reader :path
    # @return [String, nil] name alias (nil for digest-only entries)
    attr_reader :name
    # @return [String, nil] parent snapshot digest
    attr_reader :parent_digest
    # @return [String, nil] source OCI image reference
    attr_reader :image_ref
    # @return [String, nil] OCI manifest digest of the pinned image (descriptor paths)
    attr_reader :image_manifest_digest
    # @return [String, nil] payload filesystem type, e.g. "ext4" (file state only)
    attr_reader :fstype
    # @return [String, nil] best-effort source-sandbox name, if recorded
    attr_reader :source_sandbox
    # @return [Hash{String=>String}] user labels ({} for index-only entries)
    attr_reader :labels
    # @return [Integer, nil] payload size in bytes (nil for checkpoint state)
    attr_reader :size_bytes
    # @return [String, nil] stable checkpoint id (checkpoint state only)
    attr_reader :checkpoint_id
    # @return [String, nil] checkpoint-manifest digest (checkpoint state only)
    attr_reader :checkpoint_manifest_digest
    # @return [String, nil] index locality: "embedded" or "linked" (index paths)
    attr_reader :locality
    # @return [String, nil] index availability, normally "ready" (index paths)
    attr_reader :availability
    # @return [String, nil] descriptor-migration state, normally "canonical"
    #   ("reverse_complete" after downgrade tooling ran; index paths)
    attr_reader :migration_state
    # @return [String, nil] stable failure code when a v0.6.6→v0.6.7 descriptor
    #   migration was blocked (index paths)
    attr_reader :migration_error_code

    def initialize(data)
      @digest = data["digest"]
      @path = data["path"]
      @name = data["name"]
      @parent_digest = data["parent_digest"]
      @image_ref = data["image_ref"]
      @image_manifest_digest = data["image_manifest_digest"]
      @fstype = data["fstype"]
      @source_sandbox = data["source_sandbox"]
      @labels = data["labels"] || {}
      @format = data["format"]
      @size_bytes = data["size_bytes"]
      @created_at_ms = data["created_at_ms"]
      @scope = data["scope"]
      @state_kind = data["state_kind"]
      @checkpoint_id = data["checkpoint_id"]
      @checkpoint_manifest_digest = data["checkpoint_manifest_digest"]
      @locality = data["locality"]
      @availability = data["availability"]
      @migration_state = data["migration_state"]
      @migration_error_code = data["migration_error_code"]
    end

    # @return [Symbol, nil] disk format (:raw or :qcow2; nil for checkpoint state)
    def format
      @format&.to_sym
    end

    # @return [Symbol, nil] payload scope — :disk today, :resumable once VM
    #   pause/resume lands upstream
    def scope
      @scope&.to_sym
    end

    # @return [String, nil] state family: "file" or "checkpoint"
    attr_reader :state_kind

    # @return [Time, nil]
    def created_at
      @created_at_ms && Time.at(@created_at_ms / 1000.0)
    end

    # Re-open this snapshot's artifact (cheap metadata validation), returning a
    # fully-populated {SnapshotInfo}. Addresses by path, so it works even for
    # artifacts that were never added to the local index.
    # @return [SnapshotInfo]
    def open
      Snapshot.open(@path || @digest)
    end

    # Remove this snapshot's artifact and its index row.
    # @param force [Boolean] remove even if it has indexed children
    # @return [nil]
    def remove(force: false)
      Snapshot.remove(@path || @digest, force: force)
    end

    def inspect
      "#<Microsandbox::SnapshotInfo digest=#{@digest.inspect}#{" name=#{@name.inspect}" if @name}>"
    end
  end

  # The result of {Snapshot.verify}. Since runtime v0.6.9 payload integrity is
  # recorded only when the snapshot was created with `record_integrity: true`:
  # such snapshots report `:verified` (an integrity mismatch raises
  # {SnapshotIntegrityError} instead of returning), snapshots without recorded
  # integrity report `:not_recorded` with `algorithm`/`content_digest` nil.
  class SnapshotVerifyReport
    # @return [String] descriptor digest
    attr_reader :digest
    # @return [String] artifact directory path
    attr_reader :path
    # @return [Symbol] :verified or :not_recorded
    attr_reader :status
    # @return [String, nil] digest algorithm (nil when :not_recorded)
    attr_reader :algorithm
    # @return [String, nil] matched content digest (nil when :not_recorded)
    attr_reader :content_digest

    def initialize(data)
      @digest = data["digest"]
      @path = data["path"]
      @status = data["upper_status"].to_sym
      @algorithm = data["upper_algorithm"]
      @content_digest = data["upper_digest"]
    end

    # @return [Boolean] whether content integrity was recorded and matched
    def verified? = @status == :verified
  end

  # Creation and management of sandbox snapshots. A snapshot captures a stopped
  # sandbox's disk state into a portable artifact; boot from it with
  # `Sandbox.create(from_snapshot: "name-or-digest")`.
  #
  # v0.6.7 descriptor contract: artifacts are identified by their own name and
  # live at `dest_dir/<name>`; the descriptor file is `snapshot.json`. Artifacts
  # written by ≤0.10.x gems (`manifest.json`) are migrated automatically by the
  # runtime on first use — after which older gem versions can no longer read
  # them (downgrade requires the `msb self downgrade` tooling).
  class Snapshot
    class << self
      # Create a snapshot named `name` from a stopped sandbox.
      #
      # @param name [String] the snapshot's own name (its identity)
      # @param from_sandbox [String] name of the (stopped) source sandbox
      # @param dest_dir [String, nil] parent directory override — the artifact
      #   is written at `dest_dir/<name>` (default: the snapshots dir)
      # @param labels [Hash, nil] user labels
      # @param force [Boolean] overwrite an existing artifact at the destination
      # @param record_integrity [Boolean] record persistent payload integrity
      #   (a Merkle content digest) in the artifact — opt-in since runtime
      #   v0.6.9 because hashing large allocated uppers is expensive; without
      #   it {Snapshot.verify} reports `:not_recorded`
      # @param resumable [Boolean] request a resumable (memory+device) snapshot;
      #   raises {UnsupportedError} until VM pause/resume lands upstream
      # @return [SnapshotInfo]
      def create(name, from_sandbox:, dest_dir: nil, labels: nil, force: false,
        record_integrity: false, resumable: false)
        opts = {"from_sandbox" => from_sandbox.to_s}
        opts["dest_dir"] = dest_dir.to_s if dest_dir
        opts["labels"] = stringify(labels) if labels
        opts["force"] = true if force
        opts["record_integrity"] = true if record_integrity
        opts["resumable"] = true if resumable
        SnapshotInfo.new(Native::Snapshot.create(name.to_s, opts))
      end

      # Open a snapshot artifact by bare name or path (cheap metadata
      # validation; does not read the payload). Unlike {get}, this also
      # works for artifacts addressed by path that were never indexed, and it
      # returns the full descriptor.
      # @return [SnapshotInfo]
      def open(name_or_path)
        SnapshotInfo.new(Native::Snapshot.open(name_or_path.to_s))
      end

      # Metadata for a snapshot by name or digest.
      # @return [SnapshotInfo]
      def get(name_or_digest)
        SnapshotInfo.new(Native::Snapshot.get(name_or_digest.to_s))
      end

      # All snapshots indexed in the local cache.
      # @return [Array<SnapshotInfo>]
      def list
        Native::Snapshot.list.map { |info| SnapshotInfo.new(info) }
      end

      # Enumerate snapshot artifacts under a directory by parsing each
      # subdirectory's `snapshot.json`, without touching the local index — for
      # inspecting external/un-imported collections (e.g. a mounted volume).
      # @return [Array<SnapshotInfo>]
      def list_dir(dir)
        Native::Snapshot.list_dir(dir.to_s).map { |info| SnapshotInfo.new(info) }
      end

      # Rebuild the local snapshot index from a directory (defaults to the
      # configured snapshots dir). The repair for index drift or out-of-band
      # imports that {list}/{get} can't otherwise see.
      # @param dir [String, nil]
      # @return [Integer] number of indexed snapshots
      def reindex(dir = nil)
        Native::Snapshot.reindex(dir&.to_s)
      end

      # Remove a snapshot artifact by name or path.
      # @param force [Boolean] remove even if referenced
      # @return [nil]
      def remove(name_or_path, force: false)
        Native::Snapshot.remove(name_or_path.to_s, force)
        nil
      end

      # Verify a snapshot's recorded payload integrity.
      # @return [SnapshotVerifyReport]
      def verify(name_or_path)
        SnapshotVerifyReport.new(Native::Snapshot.verify(name_or_path.to_s))
      end

      # Bundle a snapshot into a `.tar.zst` (or plain `.tar`) archive.
      # Renamed from `export` in 0.11.0, mirroring the v0.6.7 SDKs.
      # @param with_parents [Boolean] include ancestor snapshots
      # @param with_image [Boolean] include OCI image artifacts (boots offline)
      # @param plain_tar [Boolean] write an uncompressed `.tar`
      # @return [nil]
      def save(name_or_path, out_path, with_parents: false, with_image: false, plain_tar: false)
        opts = {}
        opts["with_parents"] = true if with_parents
        opts["with_image"] = true if with_image
        opts["plain_tar"] = true if plain_tar
        Native::Snapshot.save(name_or_path.to_s, out_path.to_s, opts)
        nil
      end

      # Unpack a snapshot archive into the snapshots dir. Renamed from
      # `import` in 0.11.0, mirroring the v0.6.7 SDKs.
      # @param dest [String, nil] explicit destination directory
      # @return [SnapshotInfo]
      def load(archive_path, dest: nil)
        SnapshotInfo.new(Native::Snapshot.load(archive_path.to_s, dest&.to_s))
      end

      private

      def stringify(hash)
        hash.each_with_object({}) { |(k, v), acc| acc[k.to_s] = v.to_s }
      end
    end
  end
end
